Add high EMF (PWM) mode setting for Pixel 10 series

chenxiaolong opened the pull request:
Depends on: https://github.com/GrapheneOS/adevtool/pull/273

---

This adds a setting for enabling the high emission frequency (480 Hz PWM) mode available on blazer, mustang, and rango. GrapheneOS already includes the PixelDisplayService implementation. The only missing component is the setting.

In the stock OS' SettingsGoogle and PixelDisplayService, the array resource lookup for `config_availableEMValueOptions` is done dynamically. However, this resource actually exists in AOSP (https://github.com/GrapheneOS/platform_frameworks_base/commit/644b6491408f2ab5a1c9558d1c452cb747c4469b), so we can just refer to it statically like all other resources.

The two strings for this option are pulled in via adevtool, but the icon is added directly in this commit because it contains references to other resources. The icon is an exact recreation of the stock OS icon. It turns out Google just used the `visibility` icon from the Apache 2.0 licensed Material Symbols project verbatim.

https://github.com/google/material-design-icons/blob/bb04090f930e272697f2a1f0d7b352d92dfeee43/symbols/android/visibility/materialsymbolsoutlined/visibility_24px.xml

---

I primarily tested this on a Pixel Tablet. With the default build, the option is hidden as expected. If I add an overlay to specify:

```xml
<array name="config_availableEMValueOptions">
    <item>0</item>
    <item>1</item>
</array>
```

then the setting shows up and changing the setting is reflected in:

```bash
adb shell settings get secure em_value
```

Unfortunately, I won't be able to test this on my Pixel 10 Pro XL until January due to an upcoming vacation.

In case anyone wants to test this on a Pixel 10, note that PixelDisplayService does not log the `check EM_VALUE setting <value>` message to logcat when changing the setting unless you're running an eng or userdebug build.

chenxiaolong committed:
Add high EMF (PWM) mode setting for Pixel 10 series

In SettingsGoogle and PixelDisplayService, the array resource lookup for
config_availableEMValueOptions is done dynamically. However, this
resource actually exists in AOSP, so we can just refer to it statically
like all other resources.

The two strings for this option are pulled in via adevtool, but the icon
is added directly in this commit because it contains references to other
resources. The icon is an exact recreation of the stock OS icon. It
turns out Google just used the `visibility` icon from the Apache 2.0
licensed Material Symbols project verbatim.

https://github.com/google/material-design-icons/blob/bb04090f930e272697f2a1f0d7b352d92dfeee43/symbols/android/visibility/materialsymbolsoutlined/visibility_24px.xml

Signed-off-by: Andrew Gunnerson <accounts+github@chiller3.com>

pixel2user commented:
I can't use GOS on my Pixel 10 Pro till this is merged!

chenxiaolong commented:
Rebased on `16-qpr2`. I've still only tested this on a Pixel Tablet with the added overlay + `adb shell settings get secure em_value`. I won't have time to test on my Pixel 10 Pro XL for a while yet.

pixel2user commented:
@chenxiaolong - I would install a testbuild for the Pixel 10 Pro with this feature as fast as I can! But I'm not able to build one myself.

muhomorr commented on src/com/google/android/settings/display/HighEmissionFrequencyController.kt:
```diff
@@ -0,0 +1,44 @@
+        // reason we can't just use the static identifier.
+        val options = mContext.resources.getIntArray(com.android.internal.R.array.config_availableEMValueOptions)
+
+        // This differs slightly from SettingsGoogle, which checks if the array is exactly [0, 1].
```
PixelDisplayService checks whether config_availableEMValueOptions array length is >= 2

chenxiaolong replied:
Good call. I can update this to match what `SinglePixelDisplay.isPwmSwitchSupported()` does.

muhomorr commented on src/com/google/android/settings/display/HighEmissionFrequencyController.kt:
```diff
@@ -0,0 +1,44 @@
+
+class HighEmissionFrequencyController(context: Context, prefKey: String) :
+    TogglePreferenceController(context, prefKey) {
+    override fun getAvailabilityStatus(): Int {
```
This setting should be unavailable on non-Google devices

chenxiaolong replied:
Since `config_availableEMValueOptions` is technically an [AOSP option](https://cs.android.com/android/platform/superproject/+/android-latest-release:frameworks/base/core/res/res/values/config_display.xml;l=35;drc=644b6491408f2ab5a1c9558d1c452cb747c4469b), I was relying on devices not overriding the default `[0]` array to keep it hidden.

Would you prefer that I add a `"Google".equals(Build.MANUFACTURER)`-style check too?

muhomorr replied:
EM_VALUE setting is documented as follows:
```
        /**
         * Setting that holds EM_VALUE (proprietary)
         *
         * @hide
         */
        public static final String EM_VALUE =
                "em_value";
```

It can potentially have a different meaning on non-Google devices.

> Would you prefer that I add a "Google".equals(Build.MANUFACTURER)-style check too?

Yes

chenxiaolong replied:
Sure, done

chenxiaolong committed:
HighEmissionFrequencyController: Match PixelDisplayService's visibility condition

Signed-off-by: Andrew Gunnerson <accounts+github@chiller3.com>

chenxiaolong committed:
HighEmissionFrequencyController: Only enable on Google devices

em_value is documented in AOSP as "proprietary" and may have different
meaning on non-Google devices.

Signed-off-by: Andrew Gunnerson <accounts+github@chiller3.com>

pixel2user commented:
Thanks for making this feature available in GrapheneOS! 🥇 

https://grapheneos.org/releases#2026011000

ha1ix commented:
Fantastic that it is pixel 10 now, is there anyway it can be supported in older devices? The pixel 8 apparently also supports higher PWM settings [1][2]. Someone is interested in custom graphene kernel [3]

[1] https://github.com/updateing/android_kernel_google_zuma/commits/14.0.0-sultan-pwm/

[2] https://github.com/elephant-43/kernel_devices_google_shusky

[3] Post in thread 'A Mod on PWM Frequency [v3: 20241027]' https://xdaforums.com/t/a-mod-on-pwm-frequency-v3-20241027.4683727/post-90291165

Edit: I see this was mentioned previously https://github.com/GrapheneOS/os-issue-tracker/issues/3850.
